src/builders/polygons.js

```diff
@@ -7,64 +7,80 @@ import earcut from 'earcut';
+ * @param {VertexData} vertex_data The VertexData were to store the results
+ * @param {Array.<number>} vertex_template The vertex template to use
+ * @param {Object} options The texture coordinate options to apply
+ * @return {number} the number of the resulting geometries (triangles)
+ */
 export function buildPolygons (
-    polygons,
-    vertex_data, vertex_template,
-    { texcoord_index, texcoord_scale, texcoord_normalize }) {
+        polygons, vertex_data, vertex_template, { texcoord_index, texcoord_scale, texcoord_normalize }) {
 
-    var vertex_elements = vertex_data.vertex_elements;
+    let vertex_elements = vertex_data.vertex_elements,
+        element_offset = vertex_data.vertex_count,
+        num_polygons = polygons.length,
+        geom_count = 0;
 
     if (texcoord_index) {
         texcoord_normalize = texcoord_normalize || 1;
+        // Usage of "var" for function wide visibility:
         var [min_u, min_v, max_u, max_v] = texcoord_scale || default_uvs;
     }
 
-    var geom_count = 0;
-    var num_polygons = polygons.length;
-    for (var p=0; p < num_polygons; p++) {
-        var element_offset = vertex_data.vertex_count;
+    for (let p = 0; p < num_polygons; p++) {
 
-        var polygon = polygons[p];
+        let polygon = polygons[p],
+            indices = triangulatePolygon(earcut.flatten(polygon)),
+            num_indices = indices.length;
 
-        // Find polygon extents to calculate UVs, fit them to the axis-aligned bounding box
-        if (texcoord_index) {
-            var [min_x, min_y, max_x, max_y] = Geo.findBoundingBox(polygon);
-            var span_x = max_x - min_x;
-            var span_y = max_y - min_y;
-            var scale_u = (max_u - min_u) / span_x;
-            var scale_v = (max_v - min_v) / span_y;
-        }
+        // The verices and vertex-elements must not be added if earcut returns no indicies:
+        if (num_indices) {
+
+            // Find polygon extents to calculate UVs, fit them to the axis-aligned bounding box:
+            if (texcoord_index) {
+                // Usage of "var" for function wide visibility:
+                var [min_x, min_y, max_x, max_y] = Geo.findBoundingBox(polygon),
+                    span_x = max_x - min_x,
+                    span_y = max_y - min_y,
+                    scale_u = (max_u - min_u) / span_x,
+                    scale_v = (max_v - min_v) / span_y;
+            }
+
+            for (let ring_index = 0; ring_index < polygon.length; ring_index++) {
+                // Add vertex data:
+                let polygon_ring = polygon[ring_index];
+                for (let i = 0; i < polygon_ring.length; i++) {
+                    let vertex = polygon_ring[i];
+                    vertex_template[0] = vertex[0];
+                    vertex_template[1] = vertex[1];
+
+                    // Add UVs:
+                    if (texcoord_index) {
+                        vertex_template.splice(texcoord_index, 2,
```

Comment:
A note here, which may or may not be relevant any longer: back when this code was originally written, I did a lot of benchmarks to minimize geometry build time. One of the items that was important at the time was to maintain a fixed-length `vertex_template` array, and only overwrite specific values within it. By using `splice`, you're semantically removing and replacing those elements, rather than just overwriting them. This may not have an actual negative performance effect depending on how the browser handles (and could conceivably be a benefit as one mutation instead of two), but I think it should be benchmarked with a complex scene that uses UV coords to measure the average build time before/after. @meetar is that something you could do?

Reply:
:+1:

Reply:
You are right, splice is a bad choice because it creates a new array for the return value. I'll change it back to the previous approach.